staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/cbor.go

```diff
@@ -0,0 +1,111 @@
+package v1
+
+import (
+	"errors"
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+func IsMapOrNull(in []byte) bool {
+	decMode, err := cbor.DecOptions{
+		DupMapKey:           cbor.DupMapKeyQuiet,
+		FieldNameMatching:   cbor.FieldNameMatchingCaseSensitive,
+		FieldNameByteString: cbor.FieldNameByteStringAllowed,
+	}.DecMode()
+	if err != nil {
+		panic(err)
+	}
+
+	dst := new(struct{})
+	if err := decMode.Unmarshal(in, &dst); err != nil {
+		var uterr *cbor.UnmarshalTypeError
+		if errors.As(err, &uterr) && uterr.GoType == "string" {
+			return true
+		}
+		return false
+	}
+
+	return true
+}
+
+func IsArrayOrNull(in []byte) bool {
+	decMode, err := cbor.DecOptions{
+		DupMapKey:           cbor.DupMapKeyQuiet,
+		FieldNameMatching:   cbor.FieldNameMatchingCaseSensitive,
+		FieldNameByteString: cbor.FieldNameByteStringAllowed,
+	}.DecMode()
+	if err != nil {
+		panic(err)
+	}
+
+	var dst [0]any
+	if err := decMode.Unmarshal(in, &dst); err != nil {
+		var uterr *cbor.UnmarshalTypeError
+		if errors.As(err, &uterr) && uterr.GoType == "string" {
+			return true
+		}
+		return false
+	}
+
+	return true
+}
+
+func IsStringArrayOrNull(in []byte) bool {
+	decMode, err := cbor.DecOptions{
+		DupMapKey:           cbor.DupMapKeyQuiet,
+		FieldNameMatching:   cbor.FieldNameMatchingCaseSensitive,
+		FieldNameByteString: cbor.FieldNameByteStringAllowed,
+	}.DecMode()
+	if err != nil {
+		panic(err)
+	}
+
+	var dst [0]string
+	if err := decMode.Unmarshal(in, &dst); err != nil {
+		var uterr *cbor.UnmarshalTypeError
+		if errors.As(err, &uterr) {
+			if uterr.GoType == "string" {
+				return true
+			}
+		}
+		return false
+	}
+	return true
+}
+
+type cborDataItemType int
+
+const (
+	cborNullType cborDataItemType = iota
+	cborBoolType
+	cborNumberType
+	cborStringType
+	cborArrayType
+	cborMapType
+	cborTagType
+)
```

Comment:
Nit: There's technically no bool or null type. For the true, false, and null simple values, I think it's clearer and acceptably robust to look for the encoded representation of those values where you expect them.